roachtest: fix overlapping failure injection bug

dev-kyle opened the pull request:
This change fixes a bug regarding failure injections overlapping over each other when not intended. Marks the mutator step itself to also set unavailableNodes so sequential failure injections don't insert into them. Also considers `waitForStableClusterVersionStep` as incompatible when injecting a network partition recovery.

Epic: none

Release note: none

Fixes: #151522 , #151384

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151548)
<!-- Reviewable:end -->

dev-kyle committed:
roachtest: mark invalid step for network partition
The `waitForStableClusterVersionStep` cannot be run during a network
partition as the nodes require communication between each other. This
change marks it as incompatible when searching for a recovery index.

Epic: none

Release note: none

Fixes: #151384

dev-kyle committed:
roachtest: fix overlapping failure injection bug

This change fixes a bug regarding failure injections overlapping over
each other when not intended. Marks the mutator step itself to also set
unavailableNodes so sequential failure injections don't insert into
them.

Epic: none

Release note: none

Fixes: #151522

dev-kyle committed:
roachtest: add unit test for failure injection

This adds a unit test to make sure that failure injections do not
overlap, and that there is always an active failure injection when
encountering a recovery step.

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go:
```diff
@@ -496,6 +498,15 @@ func (m panicNodeMutator) Generate(
+	}
+
+	return planner.cluster.GetFailer(planner.logger, planner.cluster.CRDBNodes(), name)
+
```
super nit: stray new line :)

DarrylWong reviewed (commented):
Revert 0cda1d68b4f3147b11b6d13a89ae93bd352ae8ff ?

dev-kyle committed:
Revert "roachtest: temporarily disable network partition mutator"

This reverts commit 0cda1d68b4f3147b11b6d13a89ae93bd352ae8ff.

Dev-Kyle commented:
TFTR!

bors r=DarrylWong

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go:
```diff
@@ -964,6 +966,7 @@ func (t *Test) plan() (plan *TestPlan, retErr error) {
 			bgChans:        t.bgChans,
 			logger:         t.logger,
 			cluster:        t.cluster,
+			_getFailer:     t._getFailer,
```
>*Failer state is rather kept on the runner 

Yeah, that would be ideal since we also want want to avoid calling setup and cleanup multiple times if possible. We also want the runner to recover from any faillure injection if the test fails even if the cleanup step was never hit.

We went with the approach of passing the failer directly for simplicity (at least for the first iteration), but I don't think there was a reason other than that.

herkolategan replied:
Okay thanks, I'll attempt to add it on my PR as a separate commit.

Dev-Kyle commented:
bors retry

Dev-Kyle commented:
bors retry
